sql/wsrep_master_key_manager.cc

```diff
@@ -0,0 +1,153 @@
+    mysql_plugin_registry_release(reg_svc);
+
+    keyring_reader_service_ = nullptr;
+    keyring_writer_service_ = nullptr;
+    keyring_generator_service_ = nullptr;
+
+    logger_("Failed initializing keyring services");
+  };
+
+  if (reg_svc->acquire("keyring_reader_with_status",
+                       &h_keyring_reader_service) ||
+      reg_svc->acquire_related("keyring_writer", h_keyring_reader_service,
+                               &h_keyring_writer_service) ||
+      reg_svc->acquire_related("keyring_generator", h_keyring_reader_service,
+                               &h_keyring_generator_service)) {
+    cleanup();
+    return true;
+  }
+
+  keyring_reader_service_ =
+      reinterpret_cast<SERVICE_TYPE(keyring_reader_with_status) *>(
+          h_keyring_reader_service);
+  keyring_writer_service_ = reinterpret_cast<SERVICE_TYPE(keyring_writer) *>(
+      h_keyring_writer_service);
+  keyring_generator_service_ =
+      reinterpret_cast<SERVICE_TYPE(keyring_generator) *>(
+          h_keyring_generator_service);
+
+  mysql_plugin_registry_release(reg_svc);
+  return false;
+}
+
+void MasterKeyManager::DeInit() {
+  SERVICE_TYPE(registry) *reg_svc = mysql_plugin_registry_acquire();
+
+  if (reg_svc == nullptr) {
+    logger_("Failed to get plugin registry");
+    return;
+  }
+
+  using keyring_reader_t = SERVICE_TYPE_NO_CONST(keyring_reader_with_status);
+  using keyring_writer_t = SERVICE_TYPE_NO_CONST(keyring_writer);
+  using keyring_generator_t = SERVICE_TYPE_NO_CONST(keyring_generator);
+
+  if (keyring_reader_service_) {
+    reg_svc->release(reinterpret_cast<my_h_service>(
+        const_cast<keyring_reader_t *>(keyring_reader_service_)));
+  }
+  if (keyring_writer_service_) {
+    reg_svc->release(reinterpret_cast<my_h_service>(
+        const_cast<keyring_writer_t *>(keyring_writer_service_)));
+  }
+  if (keyring_generator_service_) {
+    reg_svc->release(reinterpret_cast<my_h_service>(
+        const_cast<keyring_generator_t *>(keyring_generator_service_)));
+  }
+
+  keyring_reader_service_ = nullptr;
+  keyring_writer_service_ = nullptr;
+  keyring_generator_service_ = nullptr;
+
+  mysql_plugin_registry_release(reg_svc);
+}
+
+bool MasterKeyManager::GenerateKey(const std::string &keyId) {
+  if (keyring_generator_service_->generate(keyId.c_str(), nullptr, key_type,
+                                           KEY_LEN) == true) {
+    return true;
+  }
+  return false;
+}
+
+std::string MasterKeyManager::GetKey(const std::string &keyId) {
+  size_t secret_length = 0;
+  size_t secret_type_length = 0;
+  my_h_keyring_reader_object reader_object = nullptr;
+  bool retval =
+      keyring_reader_service_->init(keyId.c_str(), nullptr, &reader_object);
+
+  /* Keyring error */
+  if (retval == true) {
+    logger_("Failed to initialize keyring reader service");
+    return std::string();
+  }
+  /* Key absent */
+  if (reader_object == nullptr) return std::string();
+
+  auto cleanup_guard = create_scope_guard([&] {
```

Comment:
Shouldn't this guard be somewhere earlier, like around the definition of reader_object?